typing/types.ml

```diff
@@ -135,36 +135,70 @@ and method_privacy =
   | Mprivate of field_kind
 
 (* Variance *)
+(* Variance forms a product lattice of the following partial orders:
+     0 <= may_pos <= pos
+     0 <= may_weak <= may_neg <= neg
+     0 <= inj
+   Additionally, the following implications are valid
+     pos => inj
+     neg => inj
+   Examples:
+     type 'a t        : may_pos + may_neg + may_weak
+     type 'a t = 'a   : pos
+     type 'a t = 'a -> unit : neg
+     type 'a t = ('a -> unit) -> unit : pos + may_weak
+     type 'a t = A of (('a -> unit) -> unit) : pos
+     type +'a p = ..  : may_pos + inj
+     type +!'a t      : may_pos + inj
+     type -!'a t      : may_neg + inj
+     type 'a t = A    : inj
+ *)
 
 module Variance = struct
   type t = int
   type f = May_pos | May_neg | May_weak | Inj | Pos | Neg | Inv
   let single = function
     | May_pos -> 1
-    | May_neg -> 2
+    | May_neg -> 2 + 4
     | May_weak -> 4
     | Inj -> 8
-    | Pos -> 16
-    | Neg -> 32
-    | Inv -> 64
+    | Pos -> 16 + 8 + 1
+    | Neg -> 32 + 8 + 4 + 2
+    | Inv -> 63
```

Comment:
If I understand correctly, the idea here is that we code not just one element of the partial order, but the downward closure of the element(s) of interest. (Said otherwise, an element of the lattice is a downward-closed sets of elements of the partial order.)

I would propose the following principles for readability:
- if `A <= B` (when A, B are elements of a partial order), then the code of `A` is smaller than the code of `B`; currently `May_neg, May_weak` do not respect this property.
- ideally each partial order would use consecutive bits; currently `Pos` and `May_pos` (for example) do not respect this property

Reply:
I'm not completely sure of your point. This is a partial order, so you cannot encode it faithfully in a linear one. Is there any advantage it making it topological ?
This said, I agree that it would be more natural to switch the bits for `May_weak` and `May_neg`.

Reply:
I think it makes slightly more sense if the element of each order are in the same order in the numeric representation, and also (but this is just a minor detail) if the elements of one order are placed close together in the encoding. (The proposal I guess also has the property that non-atomic elements always come after each of their atoms.)